src/v/raft/tests/heartbeats_test.cc

```diff
@@ -0,0 +1,216 @@
+
+#include <seastar/core/future.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test_log.hpp>
+#include <fmt/ostream.h>
+#include <fmt/ranges.h>
+
+#include <utility>
+#include <vector>
+
+std::vector<raft::group_id> random_group_list() {
+    const auto sz = random_generators::get_int(100, 10000);
+    raft::group_id g_id{random_generators::get_int(0, 1000)};
+    std::vector<raft::group_id> groups;
+    groups.reserve(sz);
+    for (auto i = 0; i < sz; ++i) {
+        groups.push_back(g_id);
+        g_id += raft::group_id(random_generators::get_int(1, 20));
+    }
+    return groups;
+}
+raft::heartbeat_request_v2 random_request() {
+    raft::heartbeat_request_v2 req(
+      tests::random_named_int<model::node_id>(),
+      tests::random_named_int<model::node_id>());
+    auto groups = random_group_list();
+    for (auto& g : groups) {
+        raft::group_heartbeat hb{.group = g};
+        if (tests::random_bool()) {
+            raft::heartbeat_request_data data{
+              .commit_index = tests::random_named_int<model::offset>(),
+              .term = tests::random_named_int<model::term_id>(),
+              .prev_log_index = tests::random_named_int<model::offset>(),
+              .prev_log_term = tests::random_named_int<model::term_id>(),
+              .last_visible_index = tests::random_named_int<model::offset>(),
+            };
+            hb.data = data;
+        }
+        req.add(hb);
+    }
+    return req;
+}
+
+raft::reply_result random_reply_status() {
+    return random_generators::random_choice(std::vector<raft::reply_result>{
+      raft::reply_result::success,
+      raft::reply_result::failure,
+      raft::reply_result::timeout,
+      raft::reply_result::group_unavailable});
+}
+
+raft::heartbeat_reply_v2 random_reply() {
+    raft::heartbeat_reply_v2 reply(
+      tests::random_named_int<model::node_id>(),
+      tests::random_named_int<model::node_id>());
+    auto groups = random_group_list();
+    for (auto& g : groups) {
+        if (tests::random_bool()) {
+            reply.add(
+              g,
+              random_reply_status(),
+              raft::heartbeat_reply_data{
+                .source_revision
+                = tests::random_named_int<model::revision_id>(),
+                .target_revision
+                = tests::random_named_int<model::revision_id>(),
+                .term = tests::random_named_int<model::term_id>(),
+                .last_flushed_log_index
+                = tests::random_named_int<model::offset>(),
+                .last_dirty_log_index
+                = tests::random_named_int<model::offset>(),
+                .last_term_base_offset
+                = tests::random_named_int<model::offset>(),
+              });
+        } else {
+            reply.add(g, random_reply_status());
+        }
+    }
+    return reply;
+}
+
+SEASTAR_THREAD_TEST_CASE(heartbeat_request_v2_rt) {
```

Comment:
those two types has no sync serde methods

Reply:
Probably these will be generated automatically then... saw this with the controller snapshot struct - it is easy to forget and accidentally use sync methods. Alternatively we can define sync methods that will trigger assertions so that they are not used by accident.